nanomq/apps/webhook_test.c

```diff
@@ -0,0 +1,478 @@
+	char              rest_addr[128];
+	nng_url *         url;
+	int               rv;
+
+	if ((rv = nng_mtx_alloc(&job_lock)) != 0) {
+		fatal("nng_mtx_alloc", rv);
+	}
+	job_freelist = NULL;
+
+	// Set up some strings, etc.  We use the port number
+	// from the argument list.
+	snprintf(rest_addr, sizeof(rest_addr), REST_URL, port);
+	if ((rv = nng_url_parse(&url, rest_addr)) != 0) {
+		fatal("nng_url_parse", rv);
+	}
+
+	// Create the REQ socket, and put it in raw mode, connected to
+	// the remote REP server (our inproc server in this case).
+	if ((rv = nng_req0_open(&req_sock)) != 0) {
+		fatal("nng_req0_open", rv);
+	}
+	if ((rv = nng_dial(req_sock, INPROC_URL, NULL, NNG_FLAG_NONBLOCK)) !=
+	    0) {
+		fatal("nng_dial(" INPROC_URL ")", rv);
+	}
+
+	// Get a suitable HTTP server instance.  This creates one
+	// if it doesn't already exist.
+	if ((rv = nng_http_server_hold(&server, url)) != 0) {
+		fatal("nng_http_server_hold", rv);
+	}
+
+	// Allocate the handler - we use a dynamic handler for REST
+	// using the function "rest_handle" declared above.
+	rv = nng_http_handler_alloc(&handler, url->u_path, rest_handle);
+	if (rv != 0) {
+		fatal("nng_http_handler_alloc", rv);
+	}
+
+	if ((rv = nng_http_handler_set_tree(handler)) != 0) {
+		fatal("nng_http_handler_set_tree", rv);
+	}
+
+	if ((rv = nng_http_handler_set_method(handler, NULL)) != 0) {
+		fatal("nng_http_handler_set_method", rv);
+	}
+	// We want to collect the body, and we (arbitrarily) limit this to
+	// 128KB.  The default limit is 1MB.  You can explicitly collect
+	// the data yourself with another HTTP read transaction by disabling
+	// this, but that's a lot of work, especially if you want to handle
+	// chunked transfers.
+	if ((rv = nng_http_handler_collect_body(handler, true, 1024 * 128)) !=
+	    0) {
+		fatal("nng_http_handler_collect_body", rv);
+	}
+	if ((rv = nng_http_server_add_handler(server, handler)) != 0) {
+		fatal("nng_http_handler_add_handler", rv);
+	}
+	if ((rv = nng_http_server_start(server)) != 0) {
+		printf("things go wrong\n");
+		fatal("nng_http_server_start", rv);
+	}
+
+	nng_url_free(url);
+}
+
+//
+// inproc_server - this just is a simple REP server that listens for
+// messages, and performs ROT13 on them before sending them.  This
+// doesn't have to be in the same process -- it is hear for demonstration
+// simplicity only.  (Most likely this would be somewhere else.)  Note
+// especially that this uses inproc, so nothing can get to it directly
+// from outside the process.
+//
+static void
+inproc_server(void *arg)
+{
+	nng_socket s;
+	int        rv;
+	nng_msg *  msg;
+
+	if (((rv = nng_rep0_open(&s)) != 0) ||
+	    ((rv = nng_listen(s, INPROC_URL, NULL, 0)) != 0)) {
+		fatal("unable to set up inproc", rv);
+	}
+	// This is simple enough that we don't need concurrency.  Plus it
+	// makes for an easier demo.
+	for (;;) {
+		char *body;
+		if ((rv = nng_recvmsg(s, &msg, 0)) != 0) {
+			fatal("inproc recvmsg", rv);
+		}
+		body = nng_msg_body(msg);
```

Comment:
Warning here. The variable body is unused.

Reply:
I could fix that.